openedx/core/djangoapps/content/learning_sequences/api/outlines.py

```diff
@@ -0,0 +1,323 @@
+
+User = get_user_model()
+log = logging.getLogger(__name__)
+
+# Public API...
+__all__ = [
+    'get_course_outline',
+    'get_user_course_outline',
+    'get_user_course_outline_details',
+    'replace_course_outline',
+]
+
+
+def get_course_outline(course_key: CourseKey) -> CourseOutlineData:
+    """
+    Get the outline of a course run.
+
+    There is no user-specific data or permissions applied in this function.
+
+    See the definition of CourseOutlineData for details about the data returned.
+    """
+    learning_context = _get_learning_context_for_outline(course_key)
+
+    # Check to see if it's in the cache.
+    cache_key = "learning_sequences.api.get_course_outline.v1.{}.{}".format(
+        learning_context.context_key, learning_context.published_version
+    )
+    outline_cache_result = TieredCache.get_cached_response(cache_key)
+    if outline_cache_result.is_found:
+        return outline_cache_result.value
+
+    # Fetch model data, and remember that empty Sections should still be
+    # represented (so query CourseSection explicitly instead of relying only on
+    # select_related from CourseSectionSequence).
+    section_models = CourseSection.objects \
+        .filter(learning_context=learning_context) \
+        .order_by('ordering')
+    section_sequence_models = CourseSectionSequence.objects \
+        .filter(learning_context=learning_context) \
+        .order_by('ordering') \
+        .select_related('sequence')
+
+    # Build mapping of section.id keys to sequence lists.
+    sec_ids_to_sequence_list = defaultdict(list)
+
+    for sec_seq_model in section_sequence_models:
+        sequence_model = sec_seq_model.sequence
+        sequence_data = CourseLearningSequenceData(
+            usage_key=sequence_model.usage_key,
+            title=sequence_model.title,
+            visibility=VisibilityData(
+                hide_from_toc=sec_seq_model.hide_from_toc,
+                visible_to_staff_only=sec_seq_model.visible_to_staff_only,
+            )
+        )
+        sec_ids_to_sequence_list[sec_seq_model.section_id].append(sequence_data)
+
+    sections_data = []
+    for section_model in section_models:
+        sections_data.append(
+            CourseSectionData(
+                usage_key=section_model.usage_key,
+                title=section_model.title,
+                sequences=sec_ids_to_sequence_list[section_model.id],
+                visibility=VisibilityData(
+                    hide_from_toc=section_model.hide_from_toc,
+                    visible_to_staff_only=section_model.visible_to_staff_only,
+                )
+            )
+        )
+
+    outline_data = CourseOutlineData(
+        course_key=learning_context.context_key,
+        title=learning_context.title,
+        published_at=learning_context.published_at,
+        published_version=learning_context.published_version,
+        sections=sections_data,
+    )
+    TieredCache.set_all_tiers(cache_key, outline_data, 300)
```

Comment:
Serious question: do we generally hardcode timeouts or do we create constants/settings?

Reply:
Both are found scattered around the code. I was just reluctant to put yet another setting. I don't feel very strongly about it though.